### PR TITLE
Fix kubectl misleading error

### LINKED ISSUE #57198: Misleading kubectl error: the server doesn't have a resource type "po"
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:
When using kubectl 1.9.0-beta.1+, if you run `kubectl get po`, while the API server is unreachable, you get the following misleading error: `the server doesn't have a resource type "po"`.

**What you expected to happen**:
In versions 1.9.0-alpha.3 and lower, you get the proper error: `The connection to the server 127.0.0.1:1 was refused`.

**How to reproduce it (as minimally and precisely as possible)**:
```
$ curl -sL https://storage.googleapis.com/kubernetes-release/release/v1.9.0-beta.1/bin/linux/amd64/kubectl -o /tmp/kubectl \
&& chmod +x /tmp/kubectl \
&& /tmp/kubectl -s http://127.0.0.1:1 get po 
the server doesn't have a resource type "po"
```

If you try the same command with kubectl *1.9.0-alpha.3*, you get the correct error message:
```
$ curl -sL https://storage.googleapis.com/kubernetes-release/release/v1.9.0-alpha.3/bin/linux/amd64/kubectl -o /tmp/kubectl \
&& chmod +x /tmp/kubectl \
&& /tmp/kubectl -s http://127.0.0.1:1 get po 
The connection to the server 127.0.0.1:1 was refused - did you specify the right host or port?
```
You also get the same error message in kubectl *1.9.0-beta.1*, if you use the *full resource name* instead of the abbreviation:
```
$ curl -sL https://storage.googleapis.com/kubernetes-release/release/v1.9.0-beta.1/bin/linux/amd64/kubectl -o /tmp/kubectl \
&& chmod +x /tmp/kubectl \
&& /tmp/kubectl -s http://127.0.0.1:1 get pod
The connection to the server 127.0.0.1:1 was refused - did you specify the right host or port?
```

**Anything else we need to know?**:
This is a regression between 1.9.0-alpha.3 and 1.9.0-beta.1

/sig cli

### PR DESCRIPTION
**Pull request by niuzhenguo**
**What this PR does / why we need it**:
This fixes the misleading kubectl error the server doesn't have a resource type "po" while API server is unreachable. Adds a check for "connection refused" err before fallback to hardcoded types.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #57198 

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubectl/cmd/util/factory_object_mapping.go [hunk 1]
```diff
@@ -90,6 +91,9 @@ func (f *ring1Factory) objectLoader() (meta.RESTMapper, runtime.ObjectTyper, err
 		groupResources, err = discovery.GetAPIGroupResources(discoveryClient)
 	}
 	if err != nil {
+		if strings.Contains(err.Error(), "connection refused") {
```
Comment (ericchiang) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: Error strings change. Does:

```
if _, ok := err.(net.Error); ok {
}
```

do the same thing? Can we add a unit test?

**Review by soltysh — CHANGES_REQUESTED** [review]
I still don't see the test case Eric asked for this.

**Comment by niuzhenguo** [comment]
@soltysh @ericchiang Sorry for the delay, but I don't find a proper way for testing this, could you please help here?

File: pkg/kubectl/cmd/util/factory_object_mapping.go [hunk 2]
```diff
@@ -96,6 +97,9 @@ func (f *ring1Factory) objectLoader() (meta.RESTMapper, runtime.ObjectTyper, err
 	// anything.  This mirrors old behavior that used to fallback to a hardcoded list of API types compiled
 	// into kubernetes.
 	if err != nil {
+		if _, ok := err.(net.Error); ok {
```
Comment (liggitt) [inline pkg/kubectl/cmd/util/factory_object_mapping.go]: checking for a particular error type here is extremely fragile with go's type system... wrapping errors would make this fail. Need a better way to check this, and a test exercising it if we want it to stick. I also think we'd still want to continue if we got any results (if `len(groupResources) > 0`)

**Comment by idealhack** [comment]
@niuzhenguo Is this PR still needed?

**Comment by soltysh** [comment]
The file is gone, so I'll close it. Feel free to re-open when you get to that.

**Pull request closed without merging**